sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala

```diff
@@ -49,6 +49,9 @@ trait RDG extends Expression with ExpressionWithRandomSeed {
   @transient protected lazy val seed: Long = seedExpression match {
     case e if e.dataType == IntegerType => e.eval().asInstanceOf[Int]
     case e if e.dataType == LongType => e.eval().asInstanceOf[Long]
+    case e if e.dataType == FloatType => e.eval().asInstanceOf[Float].toLong
+    case e if e.dataType == DoubleType => e.eval().asInstanceOf[Double].toLong
+    case e if e.dataType.isInstanceOf[DecimalType] => e.eval().asInstanceOf[Decimal].toLong
```

Comment:
truncation and overflow may happen here, how shall we deal it it?

Reply:
BTW do we really expect users to use float/double/decimal as the random seed?

Reply:
I checked and the existing `RAND` and `RANDN` functions only accept `IntegerType` or `LongType` for the random seed (but positive, zero, and negative values are allowed). I updated this PR to be consistent with that.